### PR TITLE
#235 Token based authentication to platform API:

### PR DESCRIPTION
**Pull request by yecli**
- Added ApiKeySecretAuth handler to support HMAC auth;
- Renamed PasswordAccessTokenHandler to UserPasswordAuthHandler;
- Added factory that creates AuthHandler based on options;
- Removed HttpClientHandler with compression for now;

### CODE REVIEW

File: VirtoCommerce.Storefront/AutoRestClients/!readme.txt [hunk 1]
```diff
@@ -11,3 +11,4 @@ Troubleshooting
 
 See AutoRest guide here:
 https://github.com/Azure/autorest/blob/master/docs/developer/guide/building-code.md#strong-name-validation-errors
+$modules = @('Subscription')
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/AutoRestClients/!readme.txt]: Undo

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 2]
```diff
@@ -43,50 +41,168 @@
...
+        /// <param name="services"></param>
+        /// <param name="setupAction"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddPollyPolicies(this IServiceCollection services, Action<PollyPoliciesOptions> setupAction = null)
+        {
+            var policyOptions = new PollyPoliciesOptions();
+            setupAction?.Invoke(policyOptions);
+
+            var policyRegistry = services.AddPolicyRegistry();
+            policyRegistry.Add(
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Multi-line blocks of code should be separated with empty line from other code.

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 3]
```diff
@@ -43,50 +41,168 @@
...
+                PolicyName.HttpRetry,
+                HttpPolicyExtensions
+                    .HandleTransientHttpError()
+                    .WaitAndRetryAsync(
+                        policyOptions.HttpRetry.Count,
+                        retryAttempt => TimeSpan.FromSeconds(Math.Pow(policyOptions.HttpRetry.BackoffPower, retryAttempt))));
+            policyRegistry.Add(
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Multi-line blocks of code should be separated with empty line from other code.

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 4]
```diff
@@ -43,50 +41,168 @@
...
+        /// <typeparam name="TModuleImplementation"></typeparam>
+        /// <typeparam name="TInnerServiceClient"></typeparam>
+        /// <param name="services"></param>
+        /// <param name="crateModuleImplementationFunc"></param>
+        /// <returns></returns>
+        public static IServiceCollection AddAutoRestClient<TIModule, TModuleImplementation, TInnerServiceClient>(this IServiceCollection services, Func<TInnerServiceClient, TModuleImplementation> crateModuleImplementationFunc)
+            where TIModule : class
+            where TModuleImplementation : class, TIModule
+            where TInnerServiceClient : ServiceClient<TInnerServiceClient>
+
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Remove

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 5]
```diff
@@ -43,50 +41,168 @@
...
+                   .AddPolicyHandlerFromRegistry(PolicyName.HttpCircuitBreaker);
+            }
+
+            services.AddSingleton<TIModule>(sp => crateModuleImplementationFunc(sp.GetRequiredService<TInnerServiceClient>()));
+
+            return services;
+        }
+
+
+
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Remove

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 6]
```diff
@@ -43,50 +41,168 @@
...
         public static void AddPlatformEndpoint(this IServiceCollection services, Action<PlatformEndpointOptions> setupAction = null)
         {
             ServicePointManager.UseNagleAlgorithm = false;
-            services.AddSingleton<VirtoCommerceApiRequestHandler>();
+            services.AddTransient<ApiKeySecretAuthHandler>();
+            services.AddTransient<UserPasswordAuthHandler>();
+            services.AddSingleton<AuthenticationHandlerFactory>();
+
+            services.AddHttpClient();
+
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Remove

File: VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs [hunk 7]
```diff
@@ -43,50 +41,168 @@
...
+
+            // TODO: Apply compression for HttpClient
+            //var httpHandlerWithCompression = new HttpClientHandler
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/DependencyInjection/ServiceCollectionExtension.cs]: Remove all commented out code

File: VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs [hunk 8]
```diff
@@ -12,19 +12,20 @@ namespace VirtoCommerce.Storefront.Infrastructure.Autorest
...
-
         private readonly PlatformEndpointOptions _options;
         private readonly IHttpClientFactory _clientFactory;
+
         /// <summary>
         /// Gets or sets the timeout
         /// </summary>
         public TimeSpan Timeout { get; set; } = TimeSpan.FromSeconds(5);
 
+
```
Comment (artem-dudarev) [inline VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs]: Remove

Comment (trueboroda) [reply VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs]: Ok. It is all in draft state.

**Comment by trueboroda** [comment]
Qoute client still generate on old version of autorest. When i try to get swager scheme i catch error. "Conflicting schemaIds: Duplicate schemaIds detected for types VirtoCommerce.Domain.Commerce.Model.Address and VirtoCommerce.QuoteModule.Web.Model.Address.

File: VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs [hunk 9]
```diff
@@ -97,21 +95,6 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
...
-                _disposed = true;
-                _lock.Dispose();
-            }
-
-            base.Dispose(disposing);
-        }
-
         private async Task<bool> RenewTokensAsync(CancellationToken cancellationToken)
         {
             if (await _lock.WaitAsync(Timeout, cancellationToken).ConfigureAwait(false))
```
Comment (yecli) [inline VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs]: Here are 2 problems:
1. Platform v2 token endpoint seems to be "/token", not "connect/token". I think need to unify with 3.0 or make it configurable.
2. If we set Endpoint Url in options like "http://localhost/admin" (without trailing slash), we are getting incorrect url for token endpoint here. Probably need to handle such situations, e.g.:
```
                    var client = _clientFactory.CreateClient();
                    var endpointUrl = _options.Url.ToString();

                    if (!endpointUrl.EndsWith('/'))
                    {
                        endpointUrl += '/';
                    }

                    var response = await client.RequestPasswordTokenAsync(new PasswordTokenRequest { Address = endpointUrl + "token", UserName = _options.UserName, Password = _options.Password });

```

Comment (yecli) [reply VirtoCommerce.Storefront/Infrastructure/Autorest/UserPasswordAuthHandler.cs]: 1. It is Ok that it works only for platform v3.
2. Fixed url part combining.

**Pull request merged**